### ISSUE TITLE
Light-Head R-CNN: In Defense of Two-Stage Object Detector

### ISSUE THREAD

**Issue by NHZlX** [NHZlX/papers#1]
https://arxiv.org/abs/1711.07264v1

**Comment by NHZlX** [comment]
<img width="817" alt="485794ac42bb814bdc5146d979a81b3e" src="https://user-images.githubusercontent.com/5595332/33364975-a605764e-d521-11e7-829a-46629f070997.png">


1. Faster RCNN  has high accuracy but high computation
2. R-FCN with a little low accuracy(relative to faster rcnn) but low computation(relative to faster rcnn).

So, a light head rcnn was combined...

**Comment by NHZlX** [comment]
# Inference demo

This is an inference demo program based on the C API of PaddlePaddle.
The demo explained here is based on the C++ code, so we need to use g++ or clang++ to compile.
The demo can be run from the command line and can be used to test the inference performance of various different models.

## Android
To compile and run this demo in an Android environment, please follow the following steps:

1. Refer to [this document](https://github.com/PaddlePaddle/Paddle/blob/develop/doc/mobile/cross_compiling_for_android_en.md) to compile the Android version of PaddlePaddle. After following the mentioned steps, make install will generate an output directory containing three subdirectories: include, lib, and third_party( `libpaddle_capi_shared.so` will be produced in the `lib` directory).

2. Compile `inference.cc` to an executable program for the Android environment as follows:
    - For armeabi-v7a
    ```
    mkdir build
    cd build

    cmake .. \
    -DANDROID_ABI=armeabi-v7a \
    -DANDROID_STANDALONE_TOOLCHAIN=your/path/to/arm_standalone_toolchain \
    -DPADDLE_ROOT=The output path generated in the first step \
    -DCMAKE_BUILD_TYPE=MinSizeRel

    make
    ```
    - For arm64-v8a
    ```
    mkdir build
    cd build

    cmake .. \
    -DANDROID_ABI=arm64-v8a \
    -DANDROID_STANDALONE_TOOLCHAIN=your/path/to/arm64_standalone_toolchain \
    -DPADDLE_ROOT=The output path generated in the first step \
    -DCMAKE_BUILD_TYPE=MinSizeRel

    make
    ```

3. Prepare a merged model.    
Models config(.py) (eg: [Mobilenet](https://github.com/PaddlePaddle/Mobile/blob/develop/models/mobilenet.py)) contain only the structure of our models. A developer can choose [model config here](https://github.com/PaddlePaddle/Mobile/tree/develop/models) to train their custom models. PaddlePaddle documentation has [several tutorials](https://github.com/PaddlePaddle/models) for building and training models. The model parameter file(.tar.gz) will be generated during the training process. There we need to merge the configuration file(.py) and the parameter file(.tar.gz) into a file. Please refer to the [details.](https://github.com/PaddlePaddle/Mobile/tree/develop/tools/merge_config_parameters)

4. Run the demo program by logging into the Android environment via [adb](https://developer.android.google.cn/studio/command-line/adb.html?hl=zh-cn#Enabling) and specifying the PaddlePaddle model from the command line as follows:
```
adb push inference libpaddle_capi_shared.so mobilenet.paddle /sdcard/test_mobilenet
adb shell
cd /sdcard/test_mobilenet
export LD_LIBRARY_PATH=./
./inference --merged_model ./mobilenet.paddle --input_size 150528
```

**Note:** `input_size` is 150528, cause that the input size is `3 * 224 * 224 = 150528`

**Comment by NHZlX** [comment]
# Merge Model

## Contents
- [Introduction](introduction)
- [Using the Merge Model Tools](using-the-merge-model-tools)
- [Optimizing Tools for Deployment](optimizing-tools-for-deployment)
	- [Reduce the merged model size](reduce-the-merged-model-size)
	- [Merge batch norms](merge-batch-norms)

	
### Introduction

When you have finished training, and want to use the C++ interface to deploy the model to the real application. We need to merge config files(.py) and parameter files(.tar.gz) into a single file.

This script is designed to take a model config(.py) and the parameters values stored in a gzip file, and output a merged model.

### Using the Merge Model Tools

User should prepare model config and model parameters first.   
We offer the link to mobilenet [config]() and [paramters]().

Merge Model can be done using the script below (and modifying the arguments According to your needs)

```
from paddle.utils.merge_model import merge_v2_model
# import your network configuration
from mobilenet import mobile_net

net = mobile_net(3 * 224 * 224, 102, 1.0)
merge_v2_model(net = net,
               param_file = './param_pass_00000.tar.gz',
               output_file = './output.paddle')
```

The arguments here are specifying the net config, where to read the parameter file, where to write the merged model to.


### Optimizing Tools for Deployment
This section provides some commonly used optimizing tools, which is very useful for mobile deployment of the model.

#### Reduce the merged model size

If we want to deploy the model to mobile, the size of the model directly affects the size of the app. So we want our model to be as small as possible. 

IOS and Android application packages are compressed before downloading.
The principle of compression is that if there are a lot of repeated numbers in the compressed file, then the compression rate will be very high. But in our model parameters, there is very low repetition, so the compression is limited. 
By the way of rounding, the approximate values in model parameters will be converted to the same value, so the parameters after round will have a lot of repetition, so compression will greatly reduce the size of our models.

Rounding parameters can be done using the script below 

```
from paddle.utils.merge_model import merge_v2_model
# import your network configuration
from mobilenet import mobile_net

net = mobile_net(3 * 224 * 224, 102, 1.0)
merge_v2_model(net = net,
               param_file = './param_pass_00000.tar.gz',
               output_file = './output.paddle'
               with_rounding = True)
```
The arguments here are specifying the net config, where to read the parameter file, where to write the merged model to, wheather to rounding parameters.

The result merged model is the same size as before, but if you run `gzip` on it, it could be about 70% smaller than the original.

#### Merge batch norms 
Many of the current models add batch norm after conv or fc layers during training. During inference, we can integrate these two operations into one operation.
We give an example of `Conv + batch norm`. 


1. Conv process   
<img width="194" alt="b850c3129cdd088e1f53b030abd3de63" src="https://user-images.githubusercontent.com/5595332/34099448-86e25496-e41a-11e7-9558-31b924a1f09f.png"> 
2. Batch norm process     
<img width="151" alt="a244a4af396f292b3eae899acada6746" src="https://user-images.githubusercontent.com/5595332/34099455-8a0914d4-e41a-11e7-8dbb-8cb42fd18d6a.png">  
<img width="139" alt="81ed60ae38f6583082d87e4f5c276bfe" src="https://user-images.githubusercontent.com/5595332/34099459-8c9818a8-e41a-11e7-9487-f801484485da.png">
3. Integrate into one process
<img width="425" alt="71bdc09682f451ead8bba5c83fdcdea0" src="https://user-images.githubusercontent.com/5595332/34099412-6c2d3080-e41a-11e7-9595-876ee8d75204.png">

Merge batch normalization can be done using the script below 


```
from paddle.utils.merge_model import merge_v2_model
# import your network configuration
from mobilenet import mobile_net

net = mobile_net(3 * 224 * 224, 102, 1.0)
merge_v2_model(net = net,
               param_file = './param_pass_00000.tar.gz',
               output_file = './output.paddle'
               merge_batch_normazlization = True)
```

Merge batch normalization speeds up the Inference by around 30%.
